### DISCUSSION TITLE
PID question: controlling ET PID based on BT background value

### DISCUSSION THREAD

**Discussion by bigl00z3** [artisan-roaster-scope/artisan#763] [Q&A] [answered]
Ok ,this one has been the next step on my side for quite a while….

I have successfully implemented a Fuji PXF PID to control my heater as described on: 
[https://www.bigloose.com/v11/index.php/home/archive/artisan-roasting-with-the-pid-fuji-pxf-and-test-of-maxwell-mta](https://www.bigloose.com/v11/index.php/home/archive/artisan-roasting-with-the-pid-fuji-pxf-and-test-of-maxwell-mta)

Most of the time I am roasting by creating an ET roast profile curve in the Artisan Designer (generally using a previous roast .alog "to improve" as the background in Designer. I then save the new profile as a.log and import is as Background).
While this works well, I would like to be able to roast and control my ET based on a BT curve instead.

I have been following the various discussions, and this one got my attention:
[https://github.com/artisan-roaster-scope/artisan/discussions/739](https://github.com/artisan-roaster-scope/artisan/discussions/739) 

In this post @MAKOMO is explaining the different historical MODBUS and Fuji PIDs implementation in Artisan. 
He is also mentioning that:
"_The Fuji PID setup can only follow the signal it is reading. This is different when using the Software PID where one can select any signal_."

In the past, I had my setup configured as a Fuji device: 
![image](https://user-images.githubusercontent.com/8886922/146360644-55c48a2f-e991-441d-b248-2c8607f5145a.png)

But I have also tried to set it as a generic MODBUS device:
![image](https://user-images.githubusercontent.com/8886922/146360785-7e666b26-609a-41a5-b361-e76d1c020277.png)

both settings are working fine with my current way of roasting.

**But now, my question is:**

**How can I use the Artisan internal PID to adjust ET based on the reading of the BT value?**

I have tried to switch the "source" from ET to BT in the PID Control Window:
![image](https://user-images.githubusercontent.com/8886922/146361269-7a0334c9-b458-4427-8ccf-47135389a5a6.png)

But this just makes my live ET curve follow the BT background curve. It doesn't adapt the ET value to have my live BT follow the Background BT. 

I obviously have missed or not understood something, but I don't know where…

Nico

update 05.09.2022 fixed url to blog as it got moved

**Comment by DB-stoutheart** [comment]
Following...very curious as well.

**Comment by hermetheuscoffee** [comment]
Hi, Nico. I use the software PID for all of my roasts, and while I know how I would do what you're asking on MY machine, it brings up 2 issues:
1. I don't know what your setup is for controlling temperature (hardware, that is).
2. This is the bigger question... WHY are you wanting an ET curve to be influenced by using BT as the "source"? This has me quite puzzled. You see, PID will simply try to make the value of your controlled device (as you're suggesting, your ET) match the value of your SV (which is the BT curve). These will never be the same. While ET and BT might both climb for the initial phase of the roast, your ET will eventually level out a bit while your BT continues to climb. 
What you should ultimately care about (at least in my opinion) is matching your ACTUAL BT to the DESIRED BT (that being the background SV curve). The ET will be whatever it needs to be to get the BT to match. For instance, on an initial cold-roaster roast, your ET curve will look different from an identical batch on a warmed up roaster. 
Am I missing something obvious that you're trying to accomplish?

**Reply by bigl00z3** [reply]
Probably my explanations that were not clear enough.

My roaster is a hot air roaster. A simplified schematic of the setup is:

![image](https://user-images.githubusercontent.com/8886922/146437219-5120e914-3c30-4b27-9b64-cf31e12637f6.png)

What in the end I would like to do (still tbc if it will work well enough to be usable) is  the following:
- Design a BT curve in the Artisan designer
- Load this designed profile in the background
- Have my roaster adjust the air temperature (before the beans in the above diagram) to follow the designed BT curve

Some helpful offline discussions helped me to better understand the different options and what I need to do. 
I will try to summarise the offline discussion below. This is not a direct solution yet, but at least to keep other users on this Discussion list of what I will try to test.

**Option 1:**
The best would need be to use the Fuji PXF ad it's regulation as it is probably better than the internal PID. To test this I will need to:

- Connect my BT probe to the Fuji PXF PID 
- Connect my ET probe to my Phidget to still have a "safety" view on my air temperature in case it goes crazy when testing.
- Run the Autotune function of the Fuji PXF to get the best P, I, D auto settings
- If required fine tune P, I, and D values
- Roast

I then need to decide if the new way of working is better than my current regulation. I then have 2 possibilities.

- Keep the new setup as is. Work by controlling the roast using BT as source for SV
- I may want to be able to run the roast the old (ET as source) and the new way (BT as source). In that case it's more costly as I will need to buy a second Fuji PXF, and add a hard switch on the front panel allowing me to use either PID's output (in my case it's a 4-20mA Out line) to the heater

**Option 2:**
I actually will test this one first as easier.
This option is to use Artisan's internal PID. For this, I will need to (still to be tested, these are the instruction I got, I will post the results once done):

- Set my BT reader, a Phidget TMP1200 (connected to my BT PT100 probe) as the main device. 
- Configure an extra device of type MODBUS to access the Fuji specifying its PV register in the MODBUS tab
- Map that extra device channel via a symbolic formula to the ET
- Define a slider which sends a MODBUS Command to the Fuji to modulate the heater 
- Finally  configure the Software PID (in the PID Control dialog via the blue PID button) to use the BT as source and that new slider as target

Once this work, based on the results I will need to decide if I still want to test Option 1 or if it's already sufficient.

I can probably test easily Option 2 one evening or over the weekend. 
For option 2 it may take a bit longer. I will need to move the roaster to access the wiring and the roaster is in my shed where right now the temperature is below freezing point... Which means I don't want to do it at night.

Nico

**Reply by hermetheuscoffee** [reply]
I now understand your setup, Nico, and we're not too far off! I also use an air loft roaster with an electric heating element controlled by a variable resistance SSR. I also use the Phidget VINT hub with the 1200 RTDs! So I fully understand what you're wanting to accomplish. But I still think your desired approach to doing so is wrong. :-)
What you actually want is for the PID to control the heat based on the DISCREPANCY between the actual BT value (from your RTD probe) and the SV of your background curve. In short, you'd like your actual BT to exactly mirror your SV curve. So let the PID control the heat of the roaster based on errors/variations between those two things! It's absolutely the best way to do it. 
What you're trying to do is so much more complicated (and unnecessarily so, I might add). You're trying to figure out "what ET values will result in the BT values that I want??" That's a tough question to answer, and the answer will change a LOT based on cold roaster vs. hot roaster, ambient air temps, batch size, bean density, etc. WAY too many variables!

Truly, what you you want is to let the PID look at the SV curve for your BT, compare that to your ACTUAL BT, then adjust the heat (quite blindly, I might add) until those two values align. I've made a very informative video about this very thing that you should check out. Hopefully by the end of the video, you'll understand why this approach makes more sense. 

https://youtu.be/5OrrdhCTBPE

-Jason

**Reply by bigl00z3** [reply]
Thanks a lot Jason, that looks great. I will definitely check your video in more details and try. This week the agenda is too full but I will get back at it asap. I will post an update below as soon as I have more data from my tests.

**Comment by bigl00z3** [comment, marked as answer]
Ok, here are the results of my tests. It took longer than planned to have a free day on the roaster. But it was definitely worth it.

I started with the mentioned above option 2 but quickly decided that I would move to option 1 as it would be a better option as the Fuji PXF will likely have a better PID algorithm.

So I:

- connected by PT100 probe to the Fuji PXF which was previously connected to the ET thermocouple type K
- The ET TCK probe was connected to the Phidget 1048.

Main configuration:
<img width="816" alt="image" src="https://user-images.githubusercontent.com/8886922/147602827-19d18bc7-a935-4274-a04e-7ad55e066c6f.png">

Modbus configuration:
<img width="691" alt="image" src="https://user-images.githubusercontent.com/8886922/147603151-6ba5390f-8b6a-4c59-bb3d-7b28598498f5.png">

And the extra devices:
<img width="1080" alt="image" src="https://user-images.githubusercontent.com/8886922/147602959-795666f4-d4b6-479a-9dbf-d842e14d2d6b.png">

Symbolisms ET/BT:
<img width="313" alt="image" src="https://user-images.githubusercontent.com/8886922/147603042-707a6c68-9725-4dc1-b79a-596751589162.png">

I had of course to change the Fuji PXF configuration from TCK to RTD (and also had to reconfigure some settings as my SV was not aligned anymore in between Artisan and the PXF).

Once connected, I ran the Autotune function of the Fuji PXF with some green beans in the roast chamber. Once I had a roughly adjusted  PID, I reloaded new green beans (using the same quantity as I generally use), started a profile and once I was mid-way, stabilised at 160°C and ran the autotune again: 

<img width="1696" alt="image" src="https://user-images.githubusercontent.com/8886922/147598868-23a00000-1f90-4376-a89b-8e21db3bac28.png">

The resulting PID settings were P: 2.7, I: 42, D: 8.1 which in Artisan are for me P: 0.27, I: 4.2 and  D: 8.1:
<img width="605" alt="image" src="https://user-images.githubusercontent.com/8886922/147603084-10ecdcb1-fa8b-4308-97c2-51e198eea27c.png">

I then loaded green beans with the new settings and started the profile:
<img width="1732" alt="image" src="https://user-images.githubusercontent.com/8886922/147602216-6a1ba2a8-a562-4661-9889-7491408c42d9.png">

There was quite a nasty ET overshoot at charge, but afterward, most of the curve was close to perfection . So I tried again with a lower charge temperature:

<img width="1719" alt="image" src="https://user-images.githubusercontent.com/8886922/147599880-f52f15b4-7afe-4cfa-ab0b-13bb73184435.png">

Which was definitely better. Now I probably need to have less sensitive P,I, D values to be applied from the "Charge" until "Turning Point". I will work on this in the next few days.

Thanks a lot to everybody for the support. I hope this can also help others who way want to do the same.

Nico

**Reply by hermetheuscoffee** [reply]
This is fantastic, Nico! You can easily apply different PID values at different points in the roast by creating buttons that use the Artisan command:
p-i-d(<p>,<i>,<d>) 
You can then either trigger them by pressing each button with unique PID values... OR you could even let some of the built-in events (charge, TP, FCs, etc.) call those commands automatically! But this is looking very, very nice as it is.
Off-topic, are you really able to run 200g batches in a Neotec? Isn't that at least a 5kg roaster?

**Reply by bigl00z3** [reply]
Yep, I know, that's in the next steps. I have used the artisan p-i-d() commande quite a bit in the past when trying to figure out. 
Now trying to find the Duty% register n° for the PXF as I am now configured in "Modbus" and not PXF, I need to call the correct register...

**Reply by bigl00z3** [reply]
And for your question which I forgot to answer, yes I can run 200g on this roaster as it is an old RFB SIgnum L (does not exist anymore). 
[https://www.yumpu.com/en/document/view/3827527/nn-15-signum-e-neuhaus-neotec](url)
200g is it's nominal capacity.

**Reply by MAKOMO** [reply]
> Now trying to find the Duty% register n° for the PXF as I am now configured in "Modbus" and not PXF, I need to call the correct register...

Could it be 2041?

**Reply by bigl00z3** [reply]
Actually I tried 41 but 2041 is also working for Duty%
It works well for positive Duty % (1-100%) but when the temperature overshoots or if it needs to go down, instead of displaying 0 it displays 6549. Don't really understand why.

**Reply by MAKOMO** [reply]
Could it be that 65490 is returned (not 6549) and that the range of this register is [-100%, 100%]? Maybe the duty range can be configured in the device like for the Artisan PID. In that case this 16bit register most likely returns a signed integer and not an unsigned as interpreted by Artisan. You can turn an unsigned integer into a 16bit signed integer by the following calculation:

```
i = 65490
s16 = (i + 2**15) % 2**16 - 2**15 = (i + 32768) % 65536 - 32768 = -46
```

Maybe Artisan should be extended to include a MODBUS decode __```sInt```__ and __```sInt32```__ next to the __```Int```__ and __```Int32```__ decoders?

For now you could use a corresponding symbolic formula to do that conversion for you or just use a max(0,x) to limit the range to the positive duty.

**Reply by bigl00z3** [reply]
Not quite sure how to use (most likely my lack of understanding):

> i = 65490
s16 = (i + 2**15) % 2**16 - 2**15 = (i + 32768) % 65536 - 32768 = -46

in the extra device tab.

When looking at the PXF communication manual for MV1 (register 41) it is mentioned:
[https://www.instrumart.com/assets/PXF-Communications-Manual.pdf](https://www.instrumart.com/assets/PXF-Communications-Manual.pdf)
![image](https://user-images.githubusercontent.com/8886922/147766699-e12fc0d8-7e86-490d-aeb4-0721a22492be.png)
This one returns a value of 6504 when the pid is not heating (cooling = 0%) and when running will display the Duty% value times 10 (needs a divided by 10).

Then for register 2041 (not documented in the above PXF communication manual).
It seems to give the correct Duty% value (no need to divide by 10) but 0% is displayed as 6549 (positive). 
I tried to use min(100,x) but whenever I am in cooling mode (at the end of a batch) it will display 100% Duty instead of 0%. Which means it will calculate my consumption as full heater power. 
What I would need  here, is a calculation that excludes anything over a value 100.

In the Artisan Help, I saw the expression part and wondered if I could use it to exclude if over 100 but not sure either:
![image](https://user-images.githubusercontent.com/8886922/147769160-c90072bc-ba34-483b-9aa1-fc5593fb2503.png)

**Reply by MAKOMO** [reply]
Using the Symbolic Formula 

`(0 if x>100 else x)`

does not work?

<img width="1056" alt="Screenshot 2022-01-05 at 19 16 22" src="https://user-images.githubusercontent.com/13909011/148268158-ba82037d-4d3e-4fcb-a38b-716663464bb2.png">

**Reply by bigl00z3** [reply]
Well of course it works! Now that you pointed it out! It is, as many times, my lack of knowledge that was the issue... LOL
Now that I see the possibilities of the Expressions, it opens the door to many other possibilities.

And once more, I am again amazed by the flexibility and the completeness of the masterpiece that Artisan is!!!

Thank you Marko

Nico

**Reply by MAKOMO** [reply]
Just added a MODBUS singed Integer "sInt16" decoder which correctly returns -0.5 if the register holds 65490 to indicate -0.5% (65336 corresponds to -2%).

**Comment by bigl00z3** [comment]
Little update on the progress.

I have tried to improve my P, I and D values for the 1st part of the roast to try to stabilise the regulation until just after turning point. 
**But it was not the right approach.** 
When charging, the BT value will always be significantly different from the SV value due to many factors (green coffee, preheating temp, batch size...). The below screen shot demonstrates well the issue. While it's probably not that bad for the green coffee, I don't like the unmanaged aesthetics of it:
![Artisan change PID](https://user-images.githubusercontent.com/8886922/148129303-eb4c6c97-0697-43fd-a2e4-34af7e0e5864.png)

The best solution is to lock the PID until just after turning point and reactivate it when the BT value was getting close to the SV value. Of course the background curve created in the designer should be adapted.
On the Fuji PXF, I do this by setting it to "Manual" mode using the command write(1,132,1) and then back to Auto mode using the command write(1,132,0). (Register 0084H which is 132 in Decimal)
![image](https://user-images.githubusercontent.com/8886922/148128851-3b878b6e-ae9b-4c4b-8aaa-01c3cbe5d6b8.png)

When doing so, the regulation is just perfect:
![Artisan Lock PID](https://user-images.githubusercontent.com/8886922/148129322-e995ad27-a63b-4e65-b607-cdca2ea62b95.jpg)

So now, my last little imperfection is my Duty% value mentioned in the post above to ensure correct CO2 calculation... but that's a minor point.

Thank you everybody for the support. I could not have done it without your help. I hope this post can also help others.

Nico 

The Blog post on my site about it (to ensure I will find it again): [https://www.bigloose.com/v11/index.php/home/archive/artisan-roasting-with-the-pid-using-bt-as-input-fuji-pxf](https://www.bigloose.com/v11/index.php/home/archive/artisan-roasting-with-the-pid-using-bt-as-input-fuji-pxf)

update 05.09.2022 fixed url to blog as it got moved
